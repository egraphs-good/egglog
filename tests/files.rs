use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
