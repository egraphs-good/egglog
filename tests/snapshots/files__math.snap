---
source: tests/files.rs
expression: serialized
---
nodes:
  String-5052067413377037545:
    op: "\"a\""
    children: []
    eclass: String-5052067413377037545
    cost: 0
  Var-5052067413377037545:
    op: Var
    children:
      - String-5052067413377037545
    eclass: "1"
    cost: 1
  Add-10912160959110460649:
    op: Add
    children:
      - Const-0
      - Sub-5996666920560749382
    eclass: "6"
    cost: 1
  Mul-8740641910606876011:
    op: Mul
    children:
      - Sub-1912573936028582372
      - Var-5052067413377037545
    eclass: "4"
    cost: 1
  Rational-0:
    op: (rational 1 1)
    children: []
    eclass: Rational-0
    cost: 0
  Const-0:
    op: Const
    children:
      - Rational-0
    eclass: "0"
    cost: 1
  Rational-5871781006564002453:
    op: (rational 2 1)
    children: []
    eclass: Rational-5871781006564002453
    cost: 0
  Const-5871781006564002453:
    op: Const
    children:
      - Rational-5871781006564002453
    eclass: "2"
    cost: 1
  Sub-1912573936028582372:
    op: Sub
    children:
      - Const-5871781006564002453
      - Const-0
    eclass: "3"
    cost: 1
  Sub-5996666920560749382:
    op: Sub
    children:
      - Var-5052067413377037545
      - Mul-8740641910606876011
    eclass: "5"
    cost: 1
root_eclasses: []
class_data:
  "1":
    type: Math
  String-5052067413377037545:
    type: String
  "6":
    type: Math
  "0":
    type: Math
  "5":
    type: Math
  "4":
    type: Math
  "3":
    type: Math
  Rational-0:
    type: Rational
  "2":
    type: Math
  Rational-5871781006564002453:
    type: Rational

